src/path.cr

```diff
@@ -0,0 +1,836 @@
+        elsif starts_with_separator?(name)
+          return name[0..0]
+        end
+      elsif absolute?(name) # posix
+        return separator_string
+      end
+    end
+
+    def drive(name)
+      if windows?
+        if !starts_with_separator?(name) && name[1]? == ':'
+          return name[0..1]
+        elsif name.starts_with?("\\\\") || name.starts_with?("//")
+          # UNC share
+          index = 0
+          last_pos = 0
+          each_part_separator_index(name) do |pos|
+            if index == 2
+              return name.byte_slice(0, pos)
+            end
+            index += 1
+            last_pos = pos
+          end
+
+          if index == 2 && last_pos < name.bytesize
+            # the entire name is a UNC share
+            return name
+          end
+        end
+      end
+
+      nil
+    end
+  end
+
+  # :nodoc:
+  NATIVE_TYPE = {% if flag?(:windows) %}
+    Type::WINDOWS
+  {% else %}
+    Type::POSIX
+  {% end %}
+
+  # Creates a new `Path` in native flavour.
+  #
+  # When compiling for a windows target, this is equal to `Path.windows()`,
+  # otherwise `Path.posix` is used.
+  def self.new(name : String = "") : Path
+    new(name.check_no_null_byte, NATIVE_TYPE)
+  end
+
+  # ditto
+  def self.new(name : String, *parts) : Path
+    new(name).join(*parts)
+  end
+
+  # ditto
+  def self.new(parts : Array | Tuple) : Path
+    new("").join(parts)
+  end
+
+  # ditto
+  def self.[](*parts) : Path
+    new("").join(parts)
+  end
+
+  # ditto
+  def self.[](path : Path) : Path
+    new(path.@name, path.@type)
+  end
+
+  # Creates a new `Path` in POSIX flavour.
+  def self.posix(name : String = "") : Path
+    new(name.check_no_null_byte, Type::POSIX)
+  end
+
+  # ditto
+  def self.posix(name : String, *parts) : Path
+    posix(name).join(parts)
+  end
+
+  # ditto
+  def self.posix(parts : Array | Tuple) : Path
+    posix("").join(parts)
+  end
+
+  # Creates a new `Path` in Windows flavour.
+  def self.windows(name : String = "") : Path
+    new(name.check_no_null_byte, Type::WINDOWS)
+  end
+
+  # ditto
+  def self.windows(name : String, *parts) : Path
+    windows(name).join(parts)
+  end
+
+  # ditto
+  def self.windows(parts : Array | Tuple) : Path
+    windows("").join(parts)
+  end
+
+  # :nodoc:
+  def initialize(@name : String, @type : Type)
+  end
+
+  # Internal helper method to create a new `Path` of the same flavour as `self`.
+  private def new_instance(string : String) : Path
+    Path.new(string, @type)
+  end
+
+  # Returns `true` if this is a Windows path.
+  def windows? : Bool
+    @type.windows?
+  end
+
+  # Returns `true` if this is a POSIX path.
+  def posix? : Bool
+    @type.posix?
+  end
+
+  # Returns `true` if this is a native path for the target platform.
+  def native? : Bool
+    @type == NATIVE_TYPE
+  end
+
+  # Returns all components of this path except the last one.
+  #
+  # ```
+  # Path["/foo/bar/file.cr"].dirname # => "/foo/bar"
+  # ```
+  def dirname : String
```

Comment:
Consider removing `dirname` and using `path.parent.to_s`?

Reply:
I think it's nice to have it, but it's not a big deal either way.